swarm/pss/prox_test.go

```diff
@@ -0,0 +1,460 @@
+	"github.com/ethereum/go-ethereum/swarm/pot"
+	"github.com/ethereum/go-ethereum/swarm/state"
+)
+
+// needed to make the enode id of the receiving node available to the handler for triggers
+type handlerContextFunc func(*testData, *adapters.NodeConfig) *handler
+
+// struct to notify reception of messages to simulation driver
+// TODO To make code cleaner:
+// - consider a separate pss unwrap to message event in sim framework (this will make eventual message propagation analysis with pss easier/possible in the future)
+// - consider also test api calls to inspect handling results of messages
+type handlerNotification struct {
+	id     enode.ID
+	serial uint64
+}
+
+type testData struct {
+	mu               sync.Mutex
+	sim              *simulation.Simulation
+	handlerDone      bool // set to true on termination of the simulation run
+	requiredMessages int
+	allowedMessages  int
+	messageCount     int
+	kademlias        map[enode.ID]*network.Kademlia
+	nodeAddrs        map[enode.ID][]byte      // make predictable overlay addresses from the generated random enode ids
+	recipients       map[int][]enode.ID       // for logging output only
+	allowed          map[int][]enode.ID       // allowed recipients
+	expectedMsgs     map[enode.ID][]uint64    // message serials we expect respective nodes to receive
+	allowedMsgs      map[enode.ID][]uint64    // message serials we expect respective nodes to receive
+	senders          map[int]enode.ID         // originating nodes of the messages (intention is to choose as far as possible from the receiving neighborhood)
+	handlerC         chan handlerNotification // passes message from pss message handler to simulation driver
+	doneC            chan struct{}            // terminates the handler channel listener
+	errC             chan error               // error to pass to main sim thread
+	msgC             chan handlerNotification // message receipt notification to main sim thread
+	msgs             [][]byte                 // recipient addresses of messages
+}
+
+var (
+	pof   = pot.DefaultPof(256) // generate messages and index them
+	topic = BytesToTopic([]byte{0xf3, 0x9e, 0x06, 0x82})
+)
+
+func (d *testData) getMsgCount() int {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return d.messageCount
+}
+
+func (d *testData) incrementMsgCount() int {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.messageCount++
+	return d.messageCount
+}
+
+func (d *testData) isDone() bool {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return d.handlerDone
+}
+
+func (d *testData) setDone() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.handlerDone = true
+}
+
+func getCmdParams(t *testing.T) (int, int) {
+	args := strings.Split(t.Name(), "/")
+	msgCount, err := strconv.ParseInt(args[2], 10, 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	nodeCount, err := strconv.ParseInt(args[1], 10, 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return int(msgCount), int(nodeCount)
+}
+
+func readSnapshot(t *testing.T, nodeCount int) simulations.Snapshot {
```

Comment:
should this not move under `swarm/network/simulation`?

Reply:
no, this is just a helper function, only relevant to this particular test, and not needed otherwise.